code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/herald.dm

```diff
@@ -0,0 +1,255 @@
+								/datum/action/innate/elite_attack/herald_teleshot,
+								/datum/action/innate/elite_attack/herald_mirror)
+								
+	var/mob/living/simple_animal/hostile/asteroid/elite/herald/mirror/my_mirror = null
+	var/is_mirror = FALSE
+								
+/mob/living/simple_animal/hostile/asteroid/elite/herald/death()
+	. = ..()
+	if(!is_mirror)
+		addtimer(CALLBACK(src, .proc/become_ghost), 8)
+	if(my_mirror != null)
+		qdel(my_mirror)
+	
+/mob/living/simple_animal/hostile/asteroid/elite/herald/proc/become_ghost()
+	icon_state = "herald_ghost"
+	return
+	
+/mob/living/simple_animal/hostile/asteroid/elite/herald/say(message, bubble_type, var/list/spans = list(), sanitize = TRUE, datum/language/language = null, ignore_spam = FALSE, forced = null)
+	. = ..()
+	playsound(get_turf(src), 'sound/magic/clockwork/invoke_general.ogg', 20, TRUE)
+	
+/datum/action/innate/elite_attack/herald_trishot
+	name = "Triple Shot"
+	button_icon_state = "herald_trishot"
+	chosen_message = "<span class='boldwarning'>You are now firing three shots in your chosen direction.</span>"
+	chosen_attack_num = 1
+	
+/datum/action/innate/elite_attack/herald_directionalshot
+	name = "Circular Shot"
+	button_icon_state = "herald_directionalshot"
+	chosen_message = "<span class='boldwarning'>You are firing projectiles in all directions.</span>"
+	chosen_attack_num = 2
+	
+/datum/action/innate/elite_attack/herald_teleshot
+	name = "Teleport Shot"
+	button_icon_state = "herald_teleshot"
+	chosen_message = "<span class='boldwarning'>You will now fire a shot which teleports you where it lands.</span>"
+	chosen_attack_num = 3
+	
+/datum/action/innate/elite_attack/herald_mirror
+	name = "Summon Mirror"
+	button_icon_state = "herald_mirror"
+	chosen_message = "<span class='boldwarning'>You will spawn a mirror which duplicates your attacks.</span>"
+	chosen_attack_num = 4
+	
+/mob/living/simple_animal/hostile/asteroid/elite/herald/OpenFire()
+	if(client)
+		switch(chosen_attack)
+			if(1)
+				herald_trishot(target)
+				if(my_mirror != null)
+					my_mirror.herald_trishot(target)
+			if(2)
+				herald_directionalshot()
+				if(my_mirror != null)
+					my_mirror.herald_directionalshot()
+			if(3)
+				herald_teleshot(target)
+				if(my_mirror != null)
+					my_mirror.herald_teleshot(target)
+			if(4)
+				herald_mirror()
+		return
+		
+	var/aiattack = rand(1,4)
+	switch(aiattack)
+		if(1)
+			herald_trishot(target)
+			if(my_mirror != null)
+				my_mirror.herald_trishot(target)
+		if(2)
+			herald_directionalshot(target)
+			if(my_mirror != null)
+				my_mirror.herald_directionalshot(target)
+		if(3)
+			herald_teleshot(target)
+			if(my_mirror != null)
+				my_mirror.herald_teleshot(target)
+		if(4)
+			herald_mirror()
+				
+/obj/projectile/herald
+	name ="death bolt"
+	icon_state= "chronobolt"
+	damage = 15
+	armour_penetration = 60
+	speed = 2
+	eyeblur = 0
+	damage_type = BRUTE
+	pass_flags = PASSTABLE
+	var/mob/living/simple_animal/hostile/hostileshooter = null
```

Comment:
Why this variable exists ? You can just use the firer from what i can see ?

Reply:
Firer is not guaranteed to be a mob, which breaks the faction check further down below.  However, I have redefined it so the variable refers to a mob now instead of specifically a hostile simple mob.